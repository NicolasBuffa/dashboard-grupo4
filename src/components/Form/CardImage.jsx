import React from "react";

function CardImage({ image }) {
  return (
    <div className="productView-container_actualImgDetail">
      <div className="picture">
        {/* <p>{image}</p> */}
        {/* <img src={image} alt="" /> */}
      </div>

      <button className="buttonProductView">Quitar </button>
    </div>
  );
}

export default CardImage;
