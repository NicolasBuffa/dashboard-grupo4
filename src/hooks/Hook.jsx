import React from 'react'

function Hook() {
  return (
    <div>Hook</div>
  )
}

export default Hook