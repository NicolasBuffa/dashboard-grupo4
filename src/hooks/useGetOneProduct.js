import { useState, useEffect } from "react";
import { useParams } from "react-router-dom";

export const useGetOneProduct = (id, initialForm = {}) => {
  const [product, setProduct] = useState(initialForm);
  const params = useParams();

  // GET ONE PRODUCT HOOK
  const getProduct = async () => {
    const resp = await fetch(`http://localhost:4000/api/product/${id}`);
    const data = await resp.json();
    setProduct(data);
  };

  useEffect(() => {
    getProduct();
    return () => {};
  }, [params.id]);

  return {
    ...product,
    product,
  };
};
