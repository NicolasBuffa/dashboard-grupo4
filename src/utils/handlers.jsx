export const updateProduct = (data) => {
  return fetch("http://localhost:4000/api/product", {
    method: "PUT",
    body: JSON.stringify(data),
    headers: {
      "Content-type": "application/json; charset=UTF-8",
    },
  });
};
export const createNreProduct = (data) => {
  return fetch("http://localhost:4000/api/product", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-type": "application/json; charset=UTF-8",
    },
  });
};
