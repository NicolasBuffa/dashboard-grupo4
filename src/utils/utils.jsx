import React from 'react'

function utils() {
  return (
    <div>OPCIONAL</div>
  )
}

export default utils