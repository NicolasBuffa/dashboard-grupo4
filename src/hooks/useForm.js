import { useState } from "react";
import { useParams } from "react-router-dom";
import { useNavigate } from "react-router-dom";
import { updateProduct } from "../utils/handlers";

export const useForm = (initialForm) => {
  const navigate = useNavigate();
  const [formState, setFormState] = useState(initialForm);
  const params = useParams();

  const onInputChange = ({ target }) => {
    const { name, value } = target;
    if (name === "price" || name === "stock") {
      setFormState({ ...formState, [name]: Number(value) });
    } else {
      setFormState({ ...formState, [name]: value });
    }
  };

  const handleIncrement = () => {
    setFormState((current) => ({ ...current, stock: formState.stock + 1 }));
  };
  const handleDecrement = () => {
    if (formState.stock >= 0) {
      setFormState({ ...formState, stock: formState.stock - 1 });
    }
  };

  const onResetForm = () => {
    setFormState(initialForm);
    navigate("/products");
  };
  const handleSubmitNewProduct = async (ev) => {
    ev.preventDefault();
    fetch("http://localhost:4000/api/product", {
      method: "POST",
      body: JSON.stringify({
        ...formState,
      }),
      headers: {
        "Content-type": "application/json; charset=UTF-8",
      },
    }).then((response) =>
      response.status === 201
        ? navigate("/products")
        : console.log(response.status)
    );
  };
  const handleSubmit = async (ev) => {
    ev.preventDefault();
    updateProduct(formState).then((response) =>
      response.status === 200
        ? navigate("/products")
        : console.log(response.status)
    );
  };
  const handdleDelete = async (ev) => {
    ev.preventDefault();
    fetch(`http://localhost:4000/api/product?id=${params.id}`, {
      method: "DELETE",
      headers: {
        "Content-type": "application/json",
      },
    }).then((response) =>
      response.status === 200
        ? navigate("/products")
        : console.log(response.status)
    );
  };
  return {
    ...formState,
    formState,
    setFormState,
    onInputChange,
    handleSubmit,
    handdleDelete,
    handleDecrement,
    handleIncrement,
    handleSubmitNewProduct,
    onResetForm,
  };
};
