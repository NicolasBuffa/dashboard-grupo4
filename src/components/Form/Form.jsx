import { useRef, useEffect } from "react";

import { useForm } from "..//../hooks/useForm";
import CardImage from "./CardImage";

function Form({ product }) {
  const refImage = useRef();
  const handdleAddImage = ({ target }) => {
    setFormState({
      ...formState,
      images: [...formState.images, refImage.current.value],
    });
  };

  let initialForm = {
    id: "",
    title: "",
    price: "",
    stock: "",
    description: "",
    category: "",
    images: [],
    rating: {
      rate: "",
      count: "",
    },
  };
  const {
    formState,
    setFormState,
    onInputChange,
    handleDecrement,
    handleIncrement,
    handleSubmit,
    onResetForm,
    price,
    description,
    title,
    stock,
  } = useForm(initialForm);

  useEffect(() => {}, [formState]);
  useEffect(() => {
    if (product.id !== null) {
      setFormState({ ...formState, ...product });
    }
  }, [product]);
  return (
    <>
      {/* Componente form */}
      <form onSubmit={handleSubmit} action="submit">
        <label>Nombre</label>
        <input
          className="inputForm"
          type="text"
          placeholder="inputValue"
          name="title"
          value={title}
          onChange={onInputChange}
        />

        <label>Valor</label>
        <input
          className="inputForm"
          type="number"
          placeholder="inputValue"
          name="price"
          min="0"
          value={price}
          onChange={onInputChange}
        />

        <label>Stock</label>
        <div>
          <button
            className="buttonStock"
            type="button"
            onClick={handleDecrement}
          >
            -
          </button>

          <input
            className="inputForm"
            type="number"
            placeholder="inputValue"
            name="stock"
            min={"0"}
            value={stock}
            onChange={(e) => {
              onInputChange(e);
            }}
          />
          <button
            className="buttonStock"
            type="button"
            onClick={handleIncrement}
          >
            +
          </button>
        </div>

        <label>Description</label>
        <textarea
          className="inputForm"
          type="text"
          placeholder="inputValue"
          name="description"
          value={description}
          onChange={onInputChange}
          rows="12"
        />

        <label>Tienda</label>
        <select className="inputForm" name="" id="">
          <option selected="selected" value="tienda1">
            Olivia Store
          </option>
          <option value="store2">Store 2</option>
          <option value="store3">Store 3</option>
          <option value="store4">Store 4</option>
        </select>
        <h3>Galeria de Imagenes</h3>

        <label>Nueva Imagen</label>
        <div className="containerUploadImages">
          <input
            className="inputForm "
            name="images"
            id="inputImg"
            type="text"
            ref={refImage}
          />
          <button
            type="button"
            className="subirImage"
            onClick={handdleAddImage}
          >
            SUBIR IMAGEN
          </button>
        </div>

        <div className="productView-container_actualImg">
          <h5>Imagenes Actuales</h5>
          {formState.images.map((image, index) => {
            return (
              <CardImage
                formState={formState}
                setFormState={setFormState}
                image={image}
                index={index}
                key={image + index}
              />
            );
          })}
        </div>
        <div className="containerSubmitButton">
          <button type="button" className="submitButton" onClick={onResetForm}>
            Cancelar
          </button>
          <button className="submitButton">Guardar </button>
        </div>
      </form>
    </>
  );
}

export default Form;
