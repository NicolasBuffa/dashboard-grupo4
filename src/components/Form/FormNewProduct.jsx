import { useRef, useEffect, useState } from "react";

import { useForm } from "../../hooks/useForm";
import CardImage from "./CardImage";

function FormNewProduct({ product }) {
  const refImage = useRef();

  const handdleAddImage = ({ target }) => {
    setFormState({
      ...formState,
      images: [...formState.images, refImage.current.value],
    });
  };

  let initialForm = {
    title: "",
    price: 0,
    stock: 0,
    description: "",
    category: "",
    images: [],
    rating: {
      rate: 0,
      count: 0,
    },
  };
  const {
    formState,
    setFormState,
    onInputChange,
    handleIncrement,
    handleDecrement,
    handleSubmitNewProduct,
    onResetForm,
    price,
    description,
    title,
    stock,
    images,
  } = useForm(initialForm);
  // useEffect(() => {
  //   if (product.id !== null) {
  //     setFormState({ ...formState, ...product });
  //   }
  // }, [product]);
  return (
    <>
      {/* Componente form */}
      <form onSubmit={handleSubmitNewProduct} action="submit">
        <label>Nombre</label>
        <input
          className="inputForm"
          type="text"
          placeholder="inputValue"
          name="title"
          value={title}
          onChange={onInputChange}
        />

        <label>Valor</label>
        <input
          className="inputForm"
          type="number"
          placeholder="inputValue"
          name="price"
          value={price}
          onChange={onInputChange}
        />

        <label>Stock</label>
        <div>
          <button
            className="buttonStock"
            type="button"
            onClick={handleDecrement}
          >
            -
          </button>

          <input
            className="inputForm"
            type="number"
            placeholder="inputValue"
            name="stock"
            value={stock}
            onChange={(e) => {
              onInputChange(e);
            }}
          />
          <button
            className="buttonStock"
            type="button"
            onClick={handleIncrement}
          >
            +
          </button>
        </div>

        <label>Description</label>
        <textarea
          className="inputForm"
          type="text"
          placeholder="inputValue"
          name="description"
          value={description}
          onChange={onInputChange}
          rows="12"
        />

        <label>Tienda</label>
        <select className="inputForm" name="" id="">
          <option value=""></option>
          <option value=""></option>
          <option value=""></option>
          <option value=""></option>
        </select>
        <h3>Galeria de Imagenes</h3>

        <label>Nueva Imagen</label>
        <div className="containerUploadImages">
          <input
            className="inputForm "
            name="images"
            id="inputImg"
            type="text"
            ref={refImage}
          />
          <button
            type="button"
            className="subirImage"
            onClick={handdleAddImage}
          >
            SUBIR IMAGEN
          </button>
        </div>

        <div className="productView-container_actualImg">
          <h5>Imagenes Actuales</h5>
          {/* {formState.images.map((image, index) => {
            return (
              <CardImage
                handleRemoveImage={handleRemoveImage}
                image={image}
                index={index}
                key={image + index}
              />
            );
          })} */}
        </div>
        <div className="containerSubmitButton">
          <button type="button" className="submitButton" onClick={onResetForm}>
            Cancelar
          </button>
          <button className="submitButton">Guardar </button>
        </div>
      </form>
      {/* Componente form fin */}
    </>
  );
}

export default FormNewProduct;
