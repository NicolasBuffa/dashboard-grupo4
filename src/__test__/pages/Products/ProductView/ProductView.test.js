import {
  act,
  render,
  screen,
  fireEvent,
  renderHook,
} from "@testing-library/react";
import Form from "../../../../components/Form/Form";
import { useForm, handleSubmit } from "../../../../hooks/useForm";
import { MemoryRouter, useHref } from "react-router-dom";
import React from "react";
import userEvent from "@testing-library/user-event";
import { updateProduct } from "..//..//..//..//utils/handlers";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => jest.fn(),
}));

jest.mock("..//..//..//..//utils/handlers");

describe("TEST vista ProductView", () => {
  let product = {
    id: "",
    title: "",
    price: "",
    stock: "",
    description: "",
    category: "",
    images: [],
    rating: {
      rate: "",
      count: "",
    },
  };
  beforeEach(() => {
    updateProduct.mockResolvedValue({
      status: 200,
    });
    render(<Form product={product} />);
  });
  test("debe de hacer match con el snpatshot", () => {
    const { container } = render(<Form product={product} />);
    expect(container).toMatchSnapshot();
  });
  test("debe hacer submit del form", async () => {
    const inputTitle = screen.getByRole("title");
    const inputStock = screen.getByRole("stock");
    const inputPrice = screen.getByRole("price");

    const form = screen.getByRole("form");
    const button = screen.getByRole("submmitButton");

    const title = "Saitama";
    const price = "3";
    const stock = "3";

    // screen.logTestingPlaygroundURL();
    fireEvent.input(inputTitle, { target: { value: title } });
    fireEvent.input(inputPrice, { target: { value: price } });
    fireEvent.input(inputStock, { target: { value: stock } });

    await act(async () => {
      await userEvent.click(button);
    });

    expect(inputTitle.value).toBe(title);
    expect(inputPrice.value).toBe(price);
    expect(inputStock.value).toBe(stock);
    expect(updateProduct).toHaveBeenCalled();
    // expect(updateProduct.mock.calls[0][0].title).toBe(title);
    // expect(updateProduct.mock.calls[0][0].stock).toBe(stock);
    // expect(updateProduct.mock.calls[0][0].price).toBe(price);
    // expect(handleSubmit()).toHaveBeenCalledTimes(1);
  });
});
