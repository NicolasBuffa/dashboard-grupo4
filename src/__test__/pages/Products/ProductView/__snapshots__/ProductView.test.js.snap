// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST vista ProductView debe de hacer match con el snpatshot 1`] = `
<div>
  <form
    action="submit"
    role="form"
  >
    <label>
      Nombre
    </label>
    <input
      class="inputForm"
      name="title"
      placeholder="inputValue"
      role="title"
      type="text"
      value=""
    />
    <label>
      Valor
    </label>
    <input
      class="inputForm"
      min="0"
      name="price"
      placeholder="inputValue"
      role="price"
      type="number"
      value=""
    />
    <label>
      Stock
    </label>
    <div>
      <button
        class="buttonStock"
        type="button"
      >
        -
      </button>
      <input
        class="inputForm"
        min="0"
        name="stock"
        placeholder="inputValue"
        role="stock"
        type="number"
        value=""
      />
      <button
        class="buttonStock"
        type="button"
      >
        +
      </button>
    </div>
    <label>
      Description
    </label>
    <textarea
      class="inputForm"
      name="description"
      placeholder="inputValue"
      rows="12"
      type="text"
    />
    <label>
      Tienda
    </label>
    <select
      class="inputForm"
      id=""
      name=""
    >
      <option
        value="tienda1"
      >
        Olivia Store
      </option>
      <option
        selected=""
        value="store2"
      >
        Store 2
      </option>
      <option
        value="store3"
      >
        Store 3
      </option>
      <option
        value="store4"
      >
        Store 4
      </option>
    </select>
    <h3>
      Galeria de Imagenes
    </h3>
    <label>
      Nueva Imagen
    </label>
    <div
      class="containerUploadImages"
    >
      <input
        class="inputForm "
        id="inputImg"
        name="images"
        type="text"
      />
      <button
        class="subirImage"
        type="button"
      >
        SUBIR IMAGEN
      </button>
    </div>
    <div
      class="productView-container_actualImg"
    >
      <h5>
        Imagenes Actuales
      </h5>
    </div>
    <div
      class="containerSubmitButton"
    >
      <button
        class="submitButton"
        type="button"
      >
        Cancelar
      </button>
      <button
        class="submitButton"
        role="submmitButton"
      >
        Guardar
         
      </button>
    </div>
  </form>
</div>
`;
